### ISSUE TITLE
Craft Cloud Compatibility? 

### ISSUE THREAD

**Issue by AdamChlan** [doublesecretagency/craft-googlemaps#121]
I'm curious if this plugin is compatible with Craft Cloud. I have a site set up on Craft Cloud, and the plugin works well when running in my local environment, but not on Craft Cloud. I am noticing a couple of the JS files the plugin is trying to call are 404.

<img width="833" alt="Screenshot 2024-10-29 at 1 57 21 PM" src="https://github.com/user-attachments/assets/c93cc3a7-4d61-4264-a7d4-50e200663447">

**Comment by lindseydiloreto** [comment]
Great question! I haven't tested with Craft Cloud yet, so I guess we're finding out now that there are some compatibility issues.

Are you doing anything unusual with AJAX or Sprig? If not, take a look at these docs on [Required JS Assets](https://plugins.doublesecretagency.com/google-maps/guides/required-js-assets/) to see if there are any helpful workarounds in there. Perhaps you could prevent the automatic loading, and manually load those files instead.

**Comment by lindseydiloreto** [comment]
I suspect Craft Cloud is having trouble loading the JS files for these two lines...

```html
<script src="https://yourwebsite.com/cpresources/[HASH]/js/googlemaps.js" defer></script>
<script src="https://yourwebsite.com/cpresources/[HASH]/js/dynamicmap.js" defer></script>
```

I'll dig in and let you know what I find. 👍

**Comment by AdamChlan** [comment]
@lindseydiloreto no, we are not doing anything unusual with AJAX or sprig on these templates. Any findings so far?

**Comment by lindseydiloreto** [comment]
I don't have a full solution yet, but this might serve as a workaround in the meantime...

1. Make a copy of `googlemaps.js` and `dynamicmap.js`, and store them directly in your project.

2. [Disable automatic loading](https://plugins.doublesecretagency.com/google-maps/guides/required-js-assets/#disable-automatic-loading) of those assets.

```twig
{{ map.tag({'assets': false}) }}
```

3. Load the assets manually, specifying **your copy** of the `googlemaps.js` and `dynamicmap.js` files...

```twig
<script src="https://maps.googleapis.com/maps/api/js?key={{ googleMaps.browserKey }}" defer></script>
<script src="https://unpkg.com/@googlemaps/markerclusterer/dist/index.min.js" defer></script>
<script src="https://yourwebsite.com/path/to/googlemaps.js" defer></script>
<script src="https://yourwebsite.com/path/to/dynamicmap.js" defer></script>
```

Hopefully that workaround does the trick until I can properly patch it for Craft Cloud support.

**Comment by AdamChlan** [comment]
Hi Lindsey. I was able to get this working with your suggestion, but just realizing that we are having issue with Craft Cloud and the server key not authenticating. I'm going to see if they can help us with that.

Also curious if any of the more recent releases resolved the issue as a whole, so I can remove the workaround.

**Comment by AdamChlan** [comment]
On further review in the control panel, when I test the API keys in the utilities area, I can see the needed JS files are 404ing. Which I suppose makes sense since the work around was only applied to the front end templates. We are a few versions behind on updates, so perhaps this has been addressed on new versions?

**Comment by timkelty** [comment]
@AdamChlan which version? JS files are working fine on Cloud for me, with latest (`doublesecretagency/craft-googlemaps:5.15`)

**Comment by AdamChlan** [comment]
Thanks for confirming @timkelty . I'm going to get my version updated and will be confirming.

**Comment by AdamChlan** [comment]
@timkelty @lindseydiloreto I have upgraded to the latest version of the plugin, and I'm still having issues on cloud. I removed the previous workaround and my front end is not loading the maps, and the backend can't do the address lookup. Here is a look at the backend.

![Image](https://github.com/user-attachments/assets/15a5c532-23df-4158-b41e-d4ebd9ad9e88)

**Comment by timkelty** [comment]
> I don't have a full solution yet, but this might serve as a workaround in the meantime...
> 
> 1. Make a copy of `googlemaps.js` and `dynamicmap.js`, and store them directly in your project.
> 2. [Disable automatic loading](https://plugins.doublesecretagency.com/google-maps/guides/required-js-assets/#disable-automatic-loading) of those assets.
> 
> {{ map.tag({'assets': false}) }}
> 3. Load the assets manually, specifying **your copy** of the `googlemaps.js` and `dynamicmap.js` files...
> 
> <script src="https://maps.googleapis.com/maps/api/js?key={{ googleMaps.browserKey }}" defer></script>
> <script src="https://unpkg.com/@googlemaps/markerclusterer/dist/index.min.js" defer></script>
> <script src="https://yourwebsite.com/path/to/googlemaps.js" defer></script>
> <script src="https://yourwebsite.com/path/to/dynamicmap.js" defer></script>
> Hopefully that workaround does the trick until I can properly patch it for Craft Cloud support.

For frontend resources, this would _not_ work, unless you used the `artifactUrl()` twig cuntion for the `googlemaps.js` and `dynamicmap.js`.

Easier than that, though, is just adding this: to your front-end template:

```
{% do view.registerAssetBundle("doublesecretagency\\googlemaps\\web\\assets\\JsApiAsset") %}
```

Note: you'll still need [this PR](https://github.com/doublesecretagency/craft-googlemaps/pull/135).


@lindseydiloreto if you make your `map.tag` Twig function can just call `\craft\web\View::registerAssetBundle`, this becomes alot simpler.

**Comment by lindseydiloreto** [comment]
I mentioned this to Adam privately but forgot to post here, Tim's PR is now available on the dev branch...

```js
"doublesecretagency/craft-googlemaps": "dev-v5-dev"
```

@timkelty Would you mind elaborating on what you meant here?...

> @lindseydiloreto if you make your map.tag Twig function can just call \craft\web\View::registerAssetBundle, this becomes alot simpler.

Just trying to get all your recommended changes in together before I cut the next release. Thanks for your insights! 🙏

**Comment by AdamChlan** [comment]
I ended up testing out the dev-v5-dev branch and confirmed that it fixed geolocation lookups in the admin area, but on the front end I still needed the following in my template to render the maps. 

```
<script src="https://maps.googleapis.com/maps/api/js?key={{ googleMaps.browserKey }}" defer></script>
<script src="https://unpkg.com/@googlemaps/markerclusterer/dist/index.min.js" defer></script>
<script src="{{ cloud.artifactUrl('/images/googlemaps.js') }}" defer></script>
<script src="{{ cloud.artifactUrl('/images/dynamicmap.js') }}" defer></script>
```

**Comment by timkelty** [comment]
@AdamChlan what does your template code look like?

With this template:

```html
<!DOCTYPE html>
<html lang="en-US">
<head>
</head>
<body>
{{ googleMaps.map().tag() }}
</body>
</html>
```

I get this, which appears to be loading the expected assets:

```html
<!DOCTYPE html>
<html lang="en-US">
<head>
    <script>window._gmData = {
        logging: true,
        cluster: [],
        infoWindows: [],
        markerCallbacks: [],
      };</script></head>
<body>
<div id="map-mjdyjb" class="gm-map" data-dna="[{&quot;type&quot;:&quot;map&quot;,&quot;locations&quot;:[],&quot;options&quot;:{&quot;id&quot;:&quot;map-mjdyjb&quot;,&quot;mapOptions&quot;:null}}]">Loading map...</div>
<script src="https://maps.googleapis.com/maps/api/js?key&amp;cache=AM93&amp;callback=Function.prototype" defer></script>
<script src="https://unpkg.com/@googlemaps/markerclusterer/dist/index.min.js" defer></script>
<script src="https://cdn.craft.cloud/xxx/builds/current/cpresources/vendor-doublesecretagency-craft-googlemaps-src-resources/js/googlemaps.js?v=1742255572" defer></script>
<script src="https://cdn.craft.cloud/xxx/builds/current/cpresources/vendor-doublesecretagency-craft-googlemaps-src-resources/js/dynamicmap.js?v=1742255572" defer></script>
<script>;
  addEventListener('load', function () {googleMaps.init('map-mjdyjb', null)});</script></body>
</html>
```

**Comment by lindseydiloreto** [comment]
@timkelty Thanks for your guidance on this! I've just released [v4.6.7](https://github.com/doublesecretagency/craft-googlemaps/releases/tag/4.6.7) and [v5.1.6](https://github.com/doublesecretagency/craft-googlemaps/releases/tag/5.1.6) which both include your recent PR. 🍺

I've also created a new thread to (at some point) address your recommendations:

- https://github.com/doublesecretagency/craft-googlemaps/issues/143

@AdamChlan Hard to say what you're bumping into at this point, I haven't had any other reports of Craft Cloud compatibility issues. If there's truly something amiss, hopefully Tim can help you get to the bottom of it.

Thank you both. I'll leave this thread open a little longer in case it does turn out to be something on my end.

**Comment by lindseydiloreto** [comment]
Seems like the dust has settled on this issue, and we can officially call it resolved. Thanks @timkelty & @AdamChlan! 👍
